io.c

```diff
@@ -9186,27 +9185,80 @@ stderr_getter(ID id, VALUE *ptr)
     return rb_ractor_stderr();
 }
 
+static VALUE
+allocate_and_open_new_file(VALUE klass)
+{
+    VALUE self = io_alloc(klass);
+    rb_io_make_open_file(self);
+    return self;
+}
+
+VALUE
+rb_io_open_descriptor(VALUE klass, int descriptor, int mode, VALUE path, VALUE timeout, struct rb_io_encoding *encoding)
+{
+
+    int state;
+    VALUE self = rb_protect(allocate_and_open_new_file, klass, &state);
+    if (state) {
+        /* if we raised an exception allocating an IO object, but the caller
+           intended to transfer ownership of this FD to us, close the fd before
+           raising the exception. Otherwise, we would leak a FD - the caller
+           expects GC to close the file, but we never got around to assigning
+           it to a rb_io. */
+        if (!(mode & FMODE_EXTERNAL)) {
+            maygvl_close(descriptor, 0);
+        }
+        rb_jump_tag(state);
+    }
+
+
+    struct rb_io *io = RFILE(self)->fptr;
+    io->self = self;
+    io->fd = descriptor;
+    io->mode = mode;
+
+    /* At this point, Ruby fully owns the descriptor, and will close it when
+       the IO gets GC'd (unless FMODE_EXTERNAL was set), no matter what happens
+       in the rest of this method. */
+
+    if (NIL_P(path)) {
+        io->pathv = Qnil;
+    }
+    else {
+        StringValue(path);
+        io->pathv = rb_str_new_frozen(path);
+    }
+
+    io->timeout = timeout;
+
+    if (encoding) {
+        io->encs = *encoding;
+    }
+
+    rb_update_max_fd(descriptor);
+
+    return self;
+}
+
 static VALUE
 prep_io(int fd, int fmode, VALUE klass, const char *path)
 {
-    rb_io_t *fp;
-    VALUE io = io_alloc(klass);
+    VALUE path_value = Qnil;
+    if (path) {
+        path_value = rb_obj_freeze(rb_str_new_cstr(path));
```

Comment:
Suuuuper nitpicky, but - if `rb_str_new_cstr` raises, this will then leak the file descriptor (because `rb_io_open_descriptor` won't have been called).

I also just realised the usages of `rb_io_open_descriptor` I added in `pty.c` have the same problem - they open a file, _then_ construct a path (in a way that can raise), and _then_ call `rb_io_open_descriptor` with said path.

I wonder if the API might be safer if we stop `rb_io_open_descriptor` from taking a path at all, and instead expose a `rb_io_set_path`?  That way the path can be constructed _after_ the FD is safely put under the management of the GC.

(or, we might decide we just don't care - probably the only way these string allocations can raise is out-of-memory conditions or very silly tracepoints?)